Make SecurityDomainAdd export an Elytron realm that uses the legacy s…

sguilhen opened the pull request:
…ecurity domain context for authentication.
- Realm is exported as a capability if the export-elytron-realm attribute is set to true in the security domain configuration.
- Handlers used by the domain sub resources have been updated to remove the additional capability before reloading the services.
- Transformer has been added to reject the new attribute if it is set to true and discard it if set to false.
- Requires PicketBox upgrade in wildfly-core (to version 5.0.0.Alpha1 or later)

sguilhen committed:
Make SecurityDomainAdd export an Elytron realm that uses the legacy security domain context for authentication.
 - Realm is exported as a capability if the export-elytron-realm attribute is set to true in the security domain configuration.
 - Handlers used by the domain sub resources have been updated to remove the security realm service before reloading the services.
 - Transformer has been added to reject the new attribute if it is set to true and discard it if set to false.
 - Requires PicketBox upgrade in wildfly-core (to version 5.0.0.Alpha1 or later)

sguilhen commented:
A quick comment about the above PR:

The legacy security domains will export an elytron compatible realm if the export-elytron-realm attribute is set to true in the domain config. Example:

``` xml
<security-domain name="Legacy" cache-type="none" export-elytron-realm="true">
    <authentication>
        ...
    </authentication>
</security-domain>
```

bstansberry commented on security/subsystem/src/main/java/org/jboss/as/security/SecurityDomainAdd.java:
```diff
@@ -156,6 +165,7 @@ public void launchServices(OperationContext context, String securityDomain, Mode
         final ApplicationPolicy applicationPolicy = createApplicationPolicy(context, securityDomain, model);
         final JSSESecurityDomain jsseSecurityDomain = createJSSESecurityDomain(context, securityDomain, model);
         final String cacheType = getAuthenticationCacheType(model);
+        final boolean exportElytronRealm = exportElytronRealm(model);
```
This should be

```
final boolean exportElytronRealm = SOME_ATTRIBUTE_DEFINITION.resolveModelAttribute(context, model).asBoolean();
```

SOME_ATTRIBUTE_DEFINITION should have a default value of false.

Your exportElytronRealm method is a non-standard impl of the same thing, and those tend to become bugs as the code evolves.

bstansberry commented on security/subsystem/src/main/java/org/jboss/as/security/SecurityExtension.java:
```diff
@@ -152,18 +155,35 @@ public void initializeParsers(ExtensionParsingContext context) {
+
+                    @Override
+                    protected boolean rejectAttribute(final PathAddress address, final String attributeName, final ModelNode attributeValue, final TransformationContext context) {
+                        return attributeValue.asBoolean();
```
You also need to reject an expression.

bstansberry replied:
Actually, you have the discard above which will discard 'false', so you can reject always. This looks like the scenario discussed here:

https://docs.jboss.org/author/display/WFLY10/Domain+Mode+Subsystem+Transformers#DomainModeSubsystemTransformers-Defaultvalueoftheattributeisthesameaslegacyimpliedbehaviour

bstansberry commented on security/subsystem/src/main/java/org/jboss/as/security/SecurityExtension.java:
```diff
@@ -226,6 +246,21 @@ private void registerTransformers_1_2_0(SubsystemRegistration subsystemRegistrat
+                    protected boolean rejectAttribute(final PathAddress address, final String attributeName, final ModelNode attributeValue, final TransformationContext context) {
+                        return attributeValue.asBoolean();
+                    }
+                }, SecurityDomainResourceDefinition.EXPORT_ELYTRON_REALM)
```
You have the discard above which will discard 'false', so you can reject always. This looks like the scenario discussed here:

https://docs.jboss.org/author/display/WFLY10/Domain+Mode+Subsystem+Transformers#DomainModeSubsystemTransformers-Defaultvalueoftheattributeisthesameaslegacyimpliedbehaviour

bstansberry commented on security/subsystem/src/main/java/org/jboss/as/security/SecurityExtension.java:
```diff
@@ -256,6 +291,31 @@ private void registerTransformers_1_2_0(SubsystemRegistration subsystemRegistrat
+                    protected boolean rejectAttribute(final PathAddress address, final String attributeName, final ModelNode attributeValue, final TransformationContext context) {
+                        return attributeValue.asBoolean();
+                    }
+                }, SecurityDomainResourceDefinition.EXPORT_ELYTRON_REALM)
```
You have the discard above which will discard 'false', so you can reject always. This looks like the scenario discussed here:

https://docs.jboss.org/author/display/WFLY10/Domain+Mode+Subsystem+Transformers#DomainModeSubsystemTransformers-Defaultvalueoftheattributeisthesameaslegacyimpliedbehaviour

bstansberry commented on security/subsystem/src/main/java/org/jboss/as/security/SecurityExtension.java:
```diff
@@ -81,7 +84,7 @@
     private static final String RESOURCE_NAME = SecurityExtension.class.getPackage().getName() + ".LocalDescriptions";
 
-    private static final ModelVersion CURRENT_MODEL_VERSION = ModelVersion.create(1, 3, 0);
+    private static final ModelVersion CURRENT_MODEL_VERSION = ModelVersion.create(1, 4, 0);
```
This should be a major version bump. 1.x is for EAP 6.

bstansberry commented on security/subsystem/src/main/resources/schema/jboss-as-security_1_3.xsd:
```diff
@@ -0,0 +1,424 @@
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           targetNamespace="urn:jboss:domain:security:1.3"
+           xmlns="urn:jboss:domain:security:1.3"
```
This should be 2.0. The 1.x series is for EAP 6.

bstansberry commented on security/subsystem/src/main/java/org/jboss/as/security/Namespace.java:
```diff
@@ -36,12 +36,13 @@
     SECURITY_1_1("urn:jboss:domain:security:1.1"),
-    SECURITY_1_2("urn:jboss:domain:security:1.2");
+    SECURITY_1_2("urn:jboss:domain:security:1.2"),
+    SECURITY_1_3("urn:jboss:domain:security:1.3");
```
2.0

sguilhen committed:
Turn export-elytron-realm into a String attribute.
 - This now allows for the specification of the name of the exported elytron realm instead of using the security domain name.
 - The security domain add/remove handlers were changed to correctly register/deregister the realm capability using the exported name.
 - The transformers have been changed to reject the attribute when it has been defined.

sguilhen commented:
Based on Brian's review I was able to turn export-elytron-realm into a String attribute that contains the name of the exported realm. Also bumped the schema to 2.0 instead of 1.x (and the model version was also bumped to 2.0.0).
I've already updated wildfly-core in the incubator to use PIcketBox 5.0.0.Alpha1, so this commit should be ok to merge without any further changes.
